File: src/main/kotlin/com/jakewharton/gitout/Benchmark.kt
package com.jakewharton.gitout

import kotlin.time.Duration
import kotlin.time.TimeSource

/**
 * A simple benchmarking utility for measuring performance of operations.
 *
 * This class provides utilities to measure execution time and track performance metrics.
 * It's designed to help identify bottlenecks and measure the impact of optimizations.
 *
 * Example usage:
 * ```
 * val benchmark = Benchmark("Repository Sync")
 * benchmark.measure("Clone operation") {
 *     syncRepository(url, destination)
 * }
 * benchmark.printSummary()
 * ```
 */
internal class Benchmark(
	private val name: String,
	private val logger: Logger? = null,
) {
	private val measurements = mutableListOf<Measurement>()
	private val startTime = TimeSource.Monotonic.markNow()

	data class Measurement(
		val operation: String,
		val duration: Duration,
		val timestamp: Duration,
	)

	/**
	 * Measures the execution time of a block of code.
	 *
	 * @param operation Name of the operation being measured
	 * @param block The code block to measure
	 * @return The result of executing the block
	 */
	inline fun <T> measure(operation: String, block: () -> T): T {
		val start = TimeSource.Monotonic.markNow()
		try {
			return block()
		} finally {
			val duration = start.elapsedNow()
			val timestamp = startTime.elapsedNow()
			measurements.add(Measurement(operation, duration, timestamp))
			logger?.debug { "BENCHMARK [$name] $operation: $duration" }
		}
	}

	/**
	 * Measures the execution time of a suspend block of code.
	 *
	 * @param operation Name of the operation being measured
	 * @param block The suspend block to measure
	 * @return The result of executing the block
	 */
	suspend inline fun <T> measureSuspend(operation: String, crossinline block: suspend () -> T): T {
		val start = TimeSource.Monotonic.markNow()
		try {
			return block()
		} finally {
			val duration = start.elapsedNow()
			val timestamp = startTime.elapsedNow()
			measurements.add(Measurement(operation, duration, timestamp))
			logger?.debug { "BENCHMARK [$name] $operation: $duration" }
		}
	}

	/**
	 * Records a timing measurement without executing code.
	 * Useful for measuring external operations.
	 */
	fun record(operation: String, duration: Duration) {
		val timestamp = startTime.elapsedNow()
		measurements.add(Measurement(operation, duration, timestamp))
		logger?.debug { "BENCHMARK [$name] $operation: $duration" }
	}

	/**
	 * Returns the total elapsed time since the benchmark was created.
	 */
	fun totalElapsed(): Duration = startTime.elapsedNow()

	/**
	 * Returns all measurements recorded so far.
	 */
	fun getMeasurements(): List<Measurement> = measurements.toList()

	/**
	 * Prints a summary of all measurements to the logger.
	 */
	fun printSummary() {
		if (measurements.isEmpty()) {
			logger?.info { "BENCHMARK [$name]: No measurements recorded" }
			return
		}

		val totalTime = totalElapsed()
		logger?.info { "BENCHMARK [$name] Summary:" }
		logger?.info { "  Total time: $totalTime" }
		logger?.info { "  Operations: ${measurements.size}" }
		logger?.info { "" }

		// Print individual measurements
		measurements.forEachIndexed { index, measurement ->
			val percentage = (measurement.duration.inWholeMilliseconds.toDouble() /
				totalTime.inWholeMilliseconds.toDouble() * 100).toInt()
			logger?.info {
				"  ${index + 1}. ${measurement.operation}: ${measurement.duration} (${percentage}%)"
			}
		}

		// Calculate and print statistics
		val durations = measurements.map { it.duration.inWholeMilliseconds }
		val sum = durations.sum()
		val avg = if (durations.isNotEmpty()) sum / durations.size else 0
		val min = durations.minOrNull() ?: 0
		val max = durations.maxOrNull() ?: 0

		logger?.info { "" }
		logger?.info { "  Statistics:" }
		logger?.info { "    Average: ${avg}ms" }
		logger?.info { "    Min: ${min}ms" }
		logger?.info { "    Max: ${max}ms" }
	}

	/**
	 * Returns a formatted string summary of the benchmark results.
	 */
	override fun toString(): String {
		val totalTime = totalElapsed()
		return buildString {
			appendLine("Benchmark: $name")
			appendLine("Total time: $totalTime")
			appendLine("Operations: ${measurements.size}")
			measurements.forEach { measurement ->
				appendLine("  - ${measurement.operation}: ${measurement.duration}")
			}
		}
	}
}

/**
 * Performance statistics for tracking sync operations.
 */
internal data class PerformanceStats(
	val totalRepositories: Int,
	val successfulSyncs: Int,
	val failedSyncs: Int,
	val totalDuration: Duration,
	val averageSyncTime: Duration,
	val fastestSync: Duration,
	val slowestSync: Duration,
	val parallelWorkers: Int,
) {
	/**
	 * Calculates the theoretical speedup from parallelization.
	 * Speedup = (total sequential time) / (total parallel time)
	 */
	fun calculateSpeedup(): Double {
		if (totalDuration.inWholeMilliseconds == 0L) return 1.0
		val sequentialTime = averageSyncTime.inWholeMilliseconds * totalRepositories
		return sequentialTime.toDouble() / totalDuration.inWholeMilliseconds.toDouble()
	}

	/**
	 * Calculates the parallel efficiency.
	 * Efficiency = Speedup / Number of Workers
	 */
	fun calculateEfficiency(): Double {
		return calculateSpeedup() / parallelWorkers
	}

	override fun toString(): String {
		return buildString {
			appendLine("Performance Statistics:")
			appendLine("  Total repositories: $totalRepositories")
			appendLine("  Successful syncs: $successfulSyncs")
			appendLine("  Failed syncs: $failedSyncs")
			appendLine("  Total duration: $totalDuration")
			appendLine("  Average sync time: $averageSyncTime")
			appendLine("  Fastest sync: $fastestSync")
			appendLine("  Slowest sync: $slowestSync")
			appendLine("  Parallel workers: $parallelWorkers")
			appendLine("  Speedup: ${"%.2f".format(calculateSpeedup())}x")
			appendLine("  Efficiency: ${"%.1f".format(calculateEfficiency() * 100)}%")
		}
	}
}

File: gitout/benchmark.py
"""Benchmarking utilities (port of Benchmark.kt).

``Benchmark`` records named operation durations (via a ``measure`` context manager or
``record``) and renders a summary. ``PerformanceStats`` computes parallel speedup and
efficiency. Durations are in milliseconds; the clock is injectable for tests.
"""

from __future__ import annotations

import time
from collections.abc import Callable, Iterator
from contextlib import contextmanager
from dataclasses import dataclass, field


@dataclass(frozen=True)
class Measurement:
    operation: str
    duration_ms: float
    timestamp_ms: float


class Benchmark:
    def __init__(self, name: str, *, clock: Callable[[], float] = time.monotonic) -> None:
        self._name = name
        self._clock = clock
        self._start = clock()
        self._measurements: list[Measurement] = []

    @contextmanager
    def measure(self, operation: str) -> Iterator[None]:
        start = self._clock()
        try:
            yield
        finally:
            self.record(operation, (self._clock() - start) * 1000)

    def record(self, operation: str, duration_ms: float) -> None:
        self._measurements.append(
            Measurement(operation, duration_ms, (self._clock() - self._start) * 1000)
        )

    def total_elapsed_ms(self) -> float:
        return (self._clock() - self._start) * 1000

    @property
    def measurements(self) -> list[Measurement]:
        return list(self._measurements)

    def summary(self) -> str:
        if not self._measurements:
            return f"BENCHMARK [{self._name}]: No measurements recorded"
        total = self.total_elapsed_ms()
        durations = [m.duration_ms for m in self._measurements]
        lines = [
            f"BENCHMARK [{self._name}] Summary:",
            f"  Total time: {total:.0f}ms",
            f"  Operations: {len(self._measurements)}",
        ]
        for index, m in enumerate(self._measurements, start=1):
            pct = int(m.duration_ms / total * 100) if total else 0
            lines.append(f"  {index}. {m.operation}: {m.duration_ms:.0f}ms ({pct}%)")
        avg = sum(durations) / len(durations)
        lines += [
            "  Statistics:",
            f"    Average: {avg:.0f}ms",
            f"    Min: {min(durations):.0f}ms",
            f"    Max: {max(durations):.0f}ms",
        ]
        return "\n".join(lines)


@dataclass(frozen=True)
class PerformanceStats:
    total_repositories: int
    successful_syncs: int
    failed_syncs: int
    total_duration_ms: float
    average_sync_time_ms: float
    fastest_sync_ms: float
    slowest_sync_ms: float
    parallel_workers: int = field(default=1)

    def calculate_speedup(self) -> float:
        """Sequential time / parallel time. Returns 1.0 when no duration recorded."""
        if self.total_duration_ms == 0:
            return 1.0
        sequential = self.average_sync_time_ms * self.total_repositories
        return sequential / self.total_duration_ms

    def calculate_efficiency(self) -> float:
        """Speedup per worker."""
        if self.parallel_workers == 0:
            return 0.0
        return self.calculate_speedup() / self.parallel_workers
